### ISSUE TITLE
Idris2: repl fails to start

### ISSUE THREAD

**Issue by mcncm** [NixOS/nixpkgs#103557]
**Describe the bug**
The Idris2 repl fails to start, instead emitting a cryptic error and immediately exiting.

**To Reproduce**
1. Install the `idris2` package, either by editing your `configuration.nix` or running `$ nix-shell -p idris2`
2. Run `$ idris2`
3. Observe the output
```
     ____    __     _         ___
    /  _/___/ /____(_)____   |__ \
    / // __  / ___/ / ___/   __/ /     Version 0.2.1
  _/ // /_/ / /  / (__  )   / __/      https://www.idris-lang.org
 /___/\__,_/_/  /_/____/   /____/      Type :? for help

Welcome to Idris 2.  Enjoy yourself!
Exception in foreign-procedure: no entry for "idris2_readBufferData"
```
before the program exits.

**Expected behavior**
I would expect to drop into a repl.

**Additional context**
I wish I knew more, and could provide more helpful context!

**Notify maintainers**

@wchresta

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.72, NixOS, 20.09.1500.edb26126d98 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(mcncm): `"home-manager-20.09"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: idris2
# a list of nixos modules affected by the problem
module:
```

**Comment by wchresta** [comment]
Hi @mcncm. Thanks for the report.

I just checked on my Ubuntu+nix and also on my NixOO 20.09 and it works fine for me:
```
$ nix-shell --pure -p '(import <nixos> {}).idris2' --run idris2
[...]
$ nix-shell --pure -p '(import <nixos-unstable> {}).idris2' --run idris2
     ____    __     _         ___                                           
    /  _/___/ /____(_)____   |__ \                                          
    / // __  / ___/ / ___/   __/ /     Version 0.2.1
  _/ // /_/ / /  / (__  )   / __/      https://www.idris-lang.org           
 /___/\__,_/_/  /_/____/   /____/      Type :? for help                     

Welcome to Idris 2.  Enjoy yourself!
Main> 
Bye for now!
```
```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.8.18, NixOS, 20.09.1784.29e9c10750e (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(root): `"nixos-20.09.1784.29e9c10750e, nixos-hardware, nixos-unstable-21.03pre250162.34ad166a830"`
 - channels(user): `"nixos-unstable-21.03pre251971.a371c107116"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Can you try using the nixos channel as above and see if you can reproduce the problem?

It looks to me like a shared library cannot be found in your case. The function that can't be found in your case is in a shared library attached via the idris2 startup script. Can you do the following?

```
$ nix-shell -p idris2
[nix-shell:~]$ which -a idris2
/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/bin/idris2

[nix-shell:~]$ cat $(which -a idris2)
#! /nix/store/2jysm3dfsgby5sw5jgj43qjrb5v79ms9-bash-4.4-p23/bin/bash -e
export CHEZ=${CHEZ-'/nix/store/kpbgiw94z1d26f1finarzkwxcgqr3qm5-chez-scheme-9.5.2/bin/scheme'}
export IDRIS2_LIBS=$IDRIS2_LIBS${IDRIS2_LIBS:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib'
export IDRIS2_DATA=$IDRIS2_DATA${IDRIS2_DATA:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/support'
export IDRIS2_PATH=$IDRIS2_PATH${IDRIS2_PATH:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/base:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/contrib:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/network:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/prelude'
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH${LD_LIBRARY_PATH:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib'
exec -a "$0" "/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/bin/.idris2-wrapped"  "$@"
```
And then for the path in `IDRIS2_LIBS` do:
```
[nix-shell:~]$ ls /nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib
libidris2_support.a  libidris2_support.so

[nix-shell:~]$ strings /nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib/libidris2_support.so | grep idris2_readBufferData
idris2_readBufferData
```

Of course, the hashes in the paths might be slightly different for you.

This path should also be the same as:
```
[nix-shell:~]$ nix repl '<nixos>'
Welcome to Nix version 2.3.7. Type :? for help.

Loading '<nixos>'...
Added 12430 variables.

nix-repl> "${idris2}"       
"/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1"
```

**Comment by mcncm** [comment]
Hi @wchresta, thanks for taking a look! Here are the outputs of my following your diagnostics literally:

```
[nix-shell:~]$ which -a idris2
/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/bin/idris2
/run/current-system/sw/bin/idris2

[nix-shell:~]$ cat $(which -a idris2)
#! /nix/store/2jysm3dfsgby5sw5jgj43qjrb5v79ms9-bash-4.4-p23/bin/bash -e
export CHEZ=${CHEZ-'/nix/store/kpbgiw94z1d26f1finarzkwxcgqr3qm5-chez-scheme-9.5.2/bin/scheme'}
export IDRIS2_LIBS=$IDRIS2_LIBS${IDRIS2_LIBS:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib'
export IDRIS2_DATA=$IDRIS2_DATA${IDRIS2_DATA:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/support'
export IDRIS2_PATH=$IDRIS2_PATH${IDRIS2_PATH:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/base:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/contrib:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/network:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/prelude'
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH${LD_LIBRARY_PATH:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib'
exec -a "$0" "/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/bin/.idris2-wrapped"  "$@"
#! /nix/store/2jysm3dfsgby5sw5jgj43qjrb5v79ms9-bash-4.4-p23/bin/bash -e
export CHEZ=${CHEZ-'/nix/store/kpbgiw94z1d26f1finarzkwxcgqr3qm5-chez-scheme-9.5.2/bin/scheme'}
export IDRIS2_LIBS=$IDRIS2_LIBS${IDRIS2_LIBS:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib'
export IDRIS2_DATA=$IDRIS2_DATA${IDRIS2_DATA:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/support'
export IDRIS2_PATH=$IDRIS2_PATH${IDRIS2_PATH:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/base:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/contrib:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/network:/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/prelude'
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH${LD_LIBRARY_PATH:+':'}'/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib'
exec -a "$0" "/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/bin/.idris2-wrapped"  "$@"



[nix-shell:~]$ ls /nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib
libidris2_support.a  libidris2_support.so

[nix-shell:~]$ strings /nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1/idris2-0.2.1/lib/libidris2_support.so | grep idris2_readBufferData
idris2_readBufferData
idris2_readBufferData


[nix-shell:~]$ nix repl '<nixos>'
Welcome to Nix version 2.3.7. Type :? for help.

Loading '<nixos>'...
Added 12439 variables.

nix-repl> "${idris2}"
"/nix/store/h53m0qp4f062mjshichwknngzjililjp-idris2-0.2.1"
```

It looks like our hashes are the same, but some of the outputs are not; _e.g._ there are _two_ instances of the string `idris2_readBufferData` in my `libidris2_support.so`, but only one in yours. That shouldn't be different, right?

The good news: It _does_ work when I use the `--pure` flag to `nix-shell`, exactly as you wrote it! This is true even on stable:
```
[nix-shell:~]$ nix-shell --pure -p '(import <nixos> {}).idris2' --run idris2
[...Idris 2 ascii art...]
Welcome to Idris 2.  Enjoy yourself!
Main> :q
Bye for now!

[nix-shell:~]$ nix-shell -p idris2 --run idris2
[...Idris 2 ascii art...]
Welcome to Idris 2.  Enjoy yourself!
Exception in foreign-procedure: no entry for "idris2_readBufferData"
```
I guess that means there's some conflict with something else in my environment? Some environment variable? I'm not really sure what it could be.

**Comment by wchresta** [comment]
Hi. Yeah it looks like something is interacting with what you have in your env. The string `idris2_readBufferData` should show up twice; seems like pasting messed it up; sorry about that.

The big hammer would be to use strace: 
```
$ nix-shell -p idris2
[nix-shell:~]$ strace idris2 2>&1 | grep ^openat
```

And then look for which files and directories it's trying to access to find the supporting shared library. For me it looks like this:
```
openat(AT_FDCWD, "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/bin/idris2", O_RDONLY) = 3
[...]
openat(AT_FDCWD, "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/bin/.idris2-wrapped", O_RDONLY) = 3
openat(AT_FDCWD, "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/lib/libidris2_support.so", O_RDONLY|O_CLOEXEC) = 4
``` 

Compare that with the `--pure` version that seems to work for you:
```
$ nix-shel --pure -p idris2
[nix-shell:~]$ strace idris2 2>&1 | grep ^openat
```

Do you have chez or scheme installed in your env?

Another thing you could try is to check which scheme is used:
```
[nix-shell:~]$ head $(which -a .idris2-wrapped)
#!/nix/store/0maz8kkzxs2z52d4h22rj1j2bavccwx1-chez-scheme-9.5.4/bin/scheme --script
chez()$>,P�30���������Z����0���������p�`���lp�f��lZb��
```

There might be a difference for your scheme in the two cases (for pure and non-pure). (You can ignore the jibberisch; only the path after #!  is interesting).


Another thing you could check is to look at some of the environment variables.

```
$ nix-shell --pure -p idris2 --run export > pure-export.log
$ nix-shell -p idris2 --run export > non-pure-export.log
```

and then compare the two outputs. Especially interesting would be differences in `CHEZ`, anything that starts with `IDRIS2_`, or `LD_LIBRARY_PATH`

**Comment by wchresta** [comment]
Also, compare the output of `idris2 --paths` in the working and non-working case:

```
idris2 --paths
+ Working Directory      :: "/home/user"
+ Source Directory       :: Nothing
+ Build Directory        :: "build"
+ Output Directory       :: "build/exec"
+ Installation Prefix    :: "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1"
+ Extra Directories      :: [".", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/base", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/contrib", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/network", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/prelude", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/prelude", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/base"]
+ CG Library Directories :: ["/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/lib", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/lib", "/home/user"]
+ Data Directories       :: ["/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/support", "/nix/store/l8rj4sn0xjhz9jysxjc65ch0y6yfpqaj-idris2-0.2.1/idris2-0.2.1/support"]
```

This might also hint to the source of the problem.

**Comment by mcncm** [comment]
Agh, I have to say, you're really good at remote-troubleshooting. Ok, here's what happened. It's totally my fault: I previously had a different OS installed, and at some point decided to try out Idris 2. In order not to pollute my /usr/bin, I manually built it somewhere under my homedir on my PATH. Later, I installed NixOS, copied over my home directory, and... Anyway, I should have realized immediately, but I didn't.

```
openat(AT_FDCWD, "/home/mcncm/.local/lib/libidris2_support.so", O_RDONLY|O_CLOEXEC) = 3
```

Thanks so much for helping to get to bottom of this. I really appreciate it!
